BrainPortal/app/models/tool_config.rb

```diff
@@ -326,22 +326,45 @@ def cbrain_task_class
 
   # Returns an array of full paths to the Singularity overlay files that
   # need to be mounted, as configured by the admin. Some of them might
-  # be patterns and will need to be resolved at run time.
+  # be patterns and will need to be resolved at run time. The dsl is
+  #      # A file located on the cluster
+  #        file:/path/to/file.squashfs
+  #      # All overlays configured for DP 1234
+  #         dp:1234
+  #      # My special overlay in CBRAIN
+  #         userfile:1234
   def singularity_overlays_full_paths
     specs = self.singularity_overlays_specs.presence
     return [] if specs.blank?
 
-    specs = specs.split(/[\s,]+/).map(&:presence).compact
+    # break onto commands based on comas, whitespace or comments
+    specs = specs.split(/([\s,]|#.*$)+/).map(&:presence).compact
     specs.map do |spec|
-      # Full path or pattern
-      next spec if spec =~ /^\// # just use that
+      # Full path or pattern (legacy)
+      next spec if spec =~ /^\// # just use that - FIXME remove after successful migration
+      cb_error "Spec #{spec}  starts with semicolon" if spec.start_with?(':')
+      knd, id_or_name = spec.split(':', 2)
+      cb_error "No semicolon in spec #{spec}" if spec.exclude? ":"
       # DP spec dp:123 or dp:name
-      id_or_name = spec.sub(/^dp:/i,"")
-      dp = DataProvider.where_id_or_name(id_or_name).first
-      cb_error "Can't find DataProvider #{id_or_name} for fetching overlays" if ! dp
-      dp_ovs = dp.singularity_overlays_full_paths rescue nil
-      cb_error "DataProvider #{id_or_name} does not have any overlays configured." if dp_ovs.blank?
-      next dp_ovs
+      case knd
+      when 'dp'
+        dp = DataProvider.where_id_or_name(id_or_name).first
+        cb_error "Can't find DataProvider #{id_or_name} for fetching overlays" if ! dp
+        dp_ovs = dp.singularity_overlays_full_paths rescue nil
+        cb_error "DataProvider #{id_or_name} does not have any overlays configured." if dp_ovs.blank?
+        next dp_ovs
+      when 'file'
+        cb_error "Provide absolute filepath in spec #{spec}." if (Pathname.new id_or_name).relative?
+        next id_or_name  # despite the variable name id_or_name for local file, id (inode) is not supported, only (full) name supported
+      when 'userfile'
+        # admin can access all files
+        userfile = Userfile.find(id_or_name) rescue nil
```

Comment:
instead of invoking the finder and raising an exception then trapping it, be gentler:

```
userfile = Userfile.where(:id => id_or_name).first
```

which will return nil if it's not there and is a lot kinder on the execution stack

Reply:
ok